src/main/java/com/arpnetworking/metrics/incubator/impl/TsdPeriodicMetricsFactory.java

```diff
@@ -0,0 +1,251 @@
+/**
+ * Copyright 2017 Inscope Metrics, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.incubator.impl;
+
+import com.arpnetworking.metrics.Counter;
+import com.arpnetworking.metrics.Metrics;
+import com.arpnetworking.metrics.MetricsFactory;
+import com.arpnetworking.metrics.Timer;
+import com.arpnetworking.metrics.Unit;
+import com.arpnetworking.metrics.impl.TsdMetricsFactory;
+import com.arpnetworking.metrics.incubator.PeriodicMetricsFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * @author Brandon Arp (brandon dot arp at inscopemetrics dot com)
+ */
+public final class TsdPeriodicMetricsFactory implements PeriodicMetricsFactory {
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void registerPolledMetric(final Consumer<Metrics> consumer) {
+        _polledMetricsRegistrations.add(consumer);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void record(final Consumer<Metrics> consumer) {
+        _currentPeriodicMetrics.readLocked(metrics -> consumer.accept(new NonClosingMetrics(metrics)));
+    }
+
+    private void cyclePeriodMetrics() {
+        _currentPeriodicMetrics.readLocked(this::recordPolledMetrics);
+        final Metrics metrics = _currentPeriodicMetrics.getAndSetReference(_metricsFactory.create());
+        metrics.close();
+    }
+
+    private void recordPolledMetrics(final Metrics metrics) {
+        for (Consumer<Metrics> polledMetric : _polledMetricsRegistrations) {
+            polledMetric.accept(metrics);
```

Comment:
This is scary because we've read locked the entire instance and given any consumer the ability to block everyone. I think there needs to be more decoupling here.